[WebXR] imported/w3c/web-platform-tests/webxr/xrWebGLLayer_constructor.https.html is randomly crashing recently

fujii opened the pull request:
#### 586f410cbf177efeb332c6ed12b38eba1efdfd39
<pre>
[WebXR] imported/w3c/web-platform-tests/webxr/xrWebGLLayer_constructor.https.html is randomly crashing recently
<a href="https://bugs.webkit.org/show_bug.cgi?id=302904">https://bugs.webkit.org/show_bug.cgi?id=302904</a>

Reviewed by Sergio Villar Senin.

An assertion failed in JSC::JSCell::classInfo() while
Inspector::createScriptCallStack was trying to get the callstack.
&gt; ASSERTION FAILED: vm().heap.mutatorState() != MutatorState::Sweeping || !vm().currentThreadIsHoldingAPILock()

WebCore shouldn&apos;t call into JSC while sweeping. ~WebXRWebGLLayer called
~WebXROpaqueFramebuffer that called
WebGLRenderingContextBase::deleteFramebuffer() that reported &quot;An opaque
framebuffer can&apos;t be deleted&quot; error with a callstack, and the assertion failed.

deleteFramebuffer() is a DOM binding which reports the error for deleting an
opaque framebuffer. ~WebXROpaqueFramebuffer shouldn&apos;t call it. Instead,
~WebGLFramebuffer automatically destroys the object properly.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer):

Canonical link: <a href="https://commits.webkit.org/303535@main">https://commits.webkit.org/303535@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a43270bf4ffe4240474476f1d5e125cfb083ebb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132749 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5244 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43825 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140278 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/84776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134619 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5519 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5003 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/101492 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/84776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135695 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3882 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/118924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82284 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83513 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/112737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/37040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142935 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4914 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/37627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/109866 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4996 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/4248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110043 "Passed tests") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/3750 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/115195 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58401 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20585 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4968 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/33543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4806 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68419 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5059 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4925 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (changes requested):
This does not sound right. The assertion is telling the right thing: You should not do something side effectful for JSC during JSC Objects' dedtructor. So it looks like it is a bug in WebCore. We should check who is calling this and need to fix it.

svillar reviewed (changes requested):
I think we cannot just leak the object. Even if we cannot call deleteframebuffer we must release m_drawFramebuffer anyway.

svillar reviewed (approved):
Actually after some more research I think the patch is correct. However we need to polish the commit message as it is not describing the issue here.

Basically we don't have to manually call deleteFramebuffer(). Whenever the m_drawFramebuffer is disposed, the WebGLFramebuffer destructor will be called. That would in turn call WebGLObject::runDestructor that will eventually call the destroyObjectImpl() implementation in WebGLFramebuffer which properly calls deleteFramebuffer. 

So we won't actually leak anything.

fujii committed:
[WebXR] imported/w3c/web-platform-tests/webxr/xrWebGLLayer_constructor.https.html is randomly crashing recently
https://bugs.webkit.org/show_bug.cgi?id=302904

Reviewed by Sergio Villar Senin.

An assertion failed in JSC::JSCell::classInfo() while
Inspector::createScriptCallStack was trying to get the callstack.
> ASSERTION FAILED: vm().heap.mutatorState() != MutatorState::Sweeping || !vm().currentThreadIsHoldingAPILock()

WebCore shouldn't call into JSC while sweeping. ~WebXRWebGLLayer called
~WebXROpaqueFramebuffer that called
WebGLRenderingContextBase::deleteFramebuffer() that reported "An opaque
framebuffer can't be deleted" error with a callstack, and the assertion failed.

deleteFramebuffer() is a DOM binding which reports the error for deleting an
opaque framebuffer. ~WebXROpaqueFramebuffer shouldn't call it. Instead,
~WebGLFramebuffer automatically destroys the object properly.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer):

Canonical link: https://commits.webkit.org/303535@main
